Script module data parallel

gqchen opened the pull request:
support data parallel for ScriptModule. 

see unit tests for testing done for this PR. I also tried traced version of resnet18 from torchvision. 

I'm yet to try a complete end-to-end data parallel training. This will be next steps.

gqchen committed:
support data parallel for ScriptModule

Summary:
support data parallel for ScriptModule. First diff.

Reviewers:

CC:

Test Plan:
 -

Revert Plan:

gqchen committed:
remove changes that's no longer needed

Summary:

Reviewers:

CC:

Test Plan:
 - php -l
 -

Revert Plan:

gqchen committed:
first complete implementation

gqchen committed:
first complete implementation

Test Plan:
to be tested

gqchen committed:
fix typos

gqchen committed:
fix one more type. Now it's able to replicate scriptmodules

gqchen committed:
fix yet another typo and add assert for nullptr

gqchen committed:
unit tests

gqchen committed:
fix lint error

gqchen commented:
just rebased with latest master.

soumith commented:
@gqchen CI seems to be failing

gqchen commented:
I am able to reproduce the building problem on Mac. Looking more into this.

gqchen committed:
fix import errors on earlier versions of Python

gqchen committed:
fix lint warnings

gqchen commented:
there are quite bit of failures in test_jit.py. Looking into that.

gqchen commented:
We have the following dependencies: 
torch.jit depends on torch/functional.py and torch.nn
torch/functional.py depends on torch.nn (this really shouldn't be the case, but anyway. I looked at the code and didn't see an obvious way to remove the dependency in none-backward-compatible way)

and with this diff, we are introducing dependency of torch.nn (as Parallel is part of torch.nn) on torch.jit. 

I am able to fix the cross dependency between torch.nn and torch.jit. However with torch/functional.py, it's not an easy fix. 

I'm going to hack something up, just to see if it fixes all issues related to this diff. But we need to find a better solution on this.

gqchen committed:
Lazy loading of torch.jit to break circular dependency

Right now there is a three way circular dependency:
torch/functiona.py depends on torch.nn
torch.nn depends on torch.jit through torch.nn.parallel (as introduced
by this PR)
torch.jit depends on torch/functiona.py

torch/functional.py really shouldn't depend on torch.nn. But there is no
obvious way to fix it in a backward-compatible way.

This is a hack to see if everything passes. And we can revisit this
circular dependency.

gqchen committed:
revert unnecessary changes

gqchen commented on torch/jit/__init__.py:
```diff
@@ -2,7 +2,14 @@
+from torch.nn.functional import upsample_nearest as functional_upsample_nearest
+from torch.nn.functional import upsample as functional_upsample
+from torch.nn.functional import upsample_bilinear as functional_upsample_bilinear
+from torch.nn.functional import assert_int_or_pair as functional_assert_int_or_pair
```
if lazy loading is fine, we can revert changes in this file as well.

gqchen commented:
@soumith @zdevito @suo if you guys can comment on the circular dependency issue, that would be great. See earlier comments for details.

gqchen committed:
revert unnecessary changes

bddppq commented:
@iotamudelta re. rocm failures
```
02:33:11 ======================================================================
02:33:11 FAIL: test_python_submodule_script (test_jit.TestDataParallel)
02:33:11 ----------------------------------------------------------------------
02:33:11 Traceback (most recent call last):
02:33:11   File "/var/lib/jenkins/workspace/test/test_jit.py", line 12804, in test_python_submodule_script
02:33:11     self.check_replicas(module, replicas)
02:33:11   File "/var/lib/jenkins/workspace/test/test_jit.py", line 12791, in check_replicas
02:33:11     self.assertEqual(replica(replica_input).data, expected_output)
02:33:11   File "/var/lib/jenkins/workspace/test/common_utils.py", line 428, in assertEqual
02:33:11     assertTensorsEqual(x, y)
02:33:11   File "/var/lib/jenkins/workspace/test/common_utils.py", line 420, in assertTensorsEqual
02:33:11     self.assertLessEqual(max_err, prec, message)
02:33:11 AssertionError: tensor(1.4169, device='cuda:1') not less than or equal to 1e-05 : 
02:33:11 
02:33:11 ======================================================================
02:33:11 FAIL: test_traced_module (test_jit.TestDataParallel)
02:33:11 ----------------------------------------------------------------------
02:33:11 Traceback (most recent call last):
02:33:11   File "/var/lib/jenkins/workspace/test/test_jit.py", line 12818, in test_traced_module
02:33:11     self.check_replicas(module, replicas)
02:33:11   File "/var/lib/jenkins/workspace/test/test_jit.py", line 12791, in check_replicas
02:33:11     self.assertEqual(replica(replica_input).data, expected_output)
02:33:11   File "/var/lib/jenkins/workspace/test/common_utils.py", line 428, in assertEqual
02:33:11     assertTensorsEqual(x, y)
02:33:11   File "/var/lib/jenkins/workspace/test/common_utils.py", line 420, in assertTensorsEqual
02:33:11     self.assertLessEqual(max_err, prec, message)
02:33:11 AssertionError: tensor(1.6529, device='cuda:1') not less than or equal to 1e-05 : 
02:33:11 
```

gqchen committed:
skip data parallel tests on AMD GPUs

Summary:
skip data parallel tests on AMD GPUs since PyTorch doesn't support
multiple AMD GPUs yet.

gqchen commented:
Two failures we have here seems to be unrelated to this PR. One is about "sccache: error : Invalid checksum", and the other is about core dump for "../.local/lib/python2.7/site-packages/caffe2/python/convnet_benchmarks_test.py::TestConvnetBenchmarks::testConvnetBenchmarks"

zdevito commented on torch/nn/parallel/replicate.py:
```diff
@@ -22,15 +104,26 @@ def replicate(network, devices, detach=False):
+                keys = set(module.__dict__.keys()) - scriptmodule_skip_attr
+                for key in keys:
+                    replica.__dict__[key] = module.__dict__[key]
+                replica.__class__ = module.__class__
```
I feel like this might be unsafe, since I am not sure that `__class__` uniquely determines the behavior of a class after it is constructed and there are 2 vs 3 differences in class behavior. Would `type(module).__new__()`, followed by `ScriptModule.__init__(replica)` work correctly? That should achieve the same result (constructing a member of the subclass, while initializing the ScriptModule appropriately), without having to mutate the `__class__` property.

gqchen replied:
this is a good idea.

gqchen replied:
I just tried the following: 

    replica = module.__new__(type(module))
    torch.jit.ScriptModule.__init__(replica)

It's coredumping. I assume it's in pybind11 related code. It's probably too hairy to go into the details there and understand why. 

Here we actually don't need to copy "__class__" attribute. That's more for my debugging purpose. I will remove this copy. The computation behavior is strictly determined by contents inside ScriptModule.

zdevito replied:
That is true. I am curious what is going one but it is almost certainly some pybind11 magic that I don't understand.

zdevito reviewed (approved):
Looks good -- I just have more minor comments.

gqchen commented on torch/nn/parallel/replicate.py:
```diff
@@ -2,9 +2,91 @@
+    if memo is None:
+        memo = set()
+
+    memo.add(module)
```
this is "visited" flag. It is memorizing that we've visited this module, whether it's legal or not. If any illegal module has been found, it will terminate early and return false. This is why in the second part we can skip any visited modules.

gqchen committed:
address review comments

gqchen committed:
add some comments for _replicatable_module

gqchen commented on torch/csrc/jit/script/module.h:
```diff
@@ -585,6 +585,22 @@ struct Module {
     }
   }
 
+  void copy_method(
```
good idea.

apaszke reviewed (commented):
Have we ran any benchmarks on this? This piece of code is very often a bottleneck in multi-GPU training and we've added a lot more of module tree traversal. Also, what about DistributedDataParallel?

gqchen commented:
@apaszke, currently DistributedDataParallel is the recommended way to do multi-GPU training, even if these GPUs are on a single machine. Long term wise, in my opinion, local-machine-multiple GPU should be handled by this DataParallel module, as it avoids network stack and in theory it can be more performant. We are pretty far away from that. This PR is the first step towards that. 

On performance of this PR,  tree traversal here (original one as well as added ones) are O(n) and it's done once per training. So I'm a bit surprised that this was a bottleneck for multi-GPU training. Do you have any benchmark you would recommend? 

 haven't done any benchmarking, other than trying this on

fmassa commented:
@gqchen I'll believe replicate is called at every invocation in forward, which could be a bottleneck in training if slowed down.

I think a representative example would be in `examples/imagenet`

gqchen commented:
@fmassa You are right. It is intended to be called on every forward, because we are doing parameter broadcasting here. Let me work on a separate PR to

gqchen commented:
It's very interesting that "replicate()" is used differently in DistributedDataParallel and DataParallel. In distributed setting (https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L237), "replicate()" is called with parameters detached and is only called during construction. We can probably do the same thing for DataParallel and move "replicate()" to be called only once.

soumith commented:
@gqchen we cant do the same for DataParallel, because of different semantics, as far as I can see.

gqchen commented:
Let me produce an exploratory PR to further this discussion.

On Tue, Feb 26, 2019 at 7:57 AM Soumith Chintala <notifications@github.com>
wrote:

> @gqchen <https://github.com/gqchen> we cant do the same for DataParallel,
> because of different semantics, as far as I can see.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/16891#issuecomment-467495561>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFfiurjVhuhLwxVjrUDMbq1KV_p8ukkiks5vRVl0gaJpZM4auRHC>
> .
>

gqchen commented:
@soumith @apaszke here is an exploratory PR: https://github.com/pytorch/pytorch/pull/17540. Please let me know at high level this makes sense. That code is a quick hack.

ziyuv commented:
Is there a plan to support this feature for the C++ frontend?

suo commented:
cc @yf225
